.github/workflows/build.yml

```diff
@@ -0,0 +1,124 @@
+name: Wheel Builder
+
+on:
+  push:
+    branches: ["main"]
+  pull_request:
+    types: [labeled, opened, synchronize, reopened]
+  release:
+    types: [published]
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+permissions:
+  contents: read # to fetch code (actions/checkout)
+
+jobs:
+  build_wheels_sdist:
+    name: Build and Test ${{ github.repository }} Pure Python Wheels
+    runs-on: ubuntu-latest
+    strategy:
+      # Ensure that a wheel builder finishes even if another fails
+      fail-fast: false
+      matrix:
+        python-version: [3.8]
+        poetry-version: [latest]
+
+    steps:
+      - name: Checkout ${{ github.repository }}
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+
+      - name: Install Poetry
+        uses: snok/install-poetry@v1
+        with:
+          version: ${{matrix.poetry-version}}
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+          installer-parallel: true
+
+      - name: Set up Python ${{matrix.python-version}}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{matrix.python-version}}
+          cache: "poetry"
+
+
+      - name: Linting, Typechecking and Formatting
+        uses: pre-commit/action@v3.0.0
+
+      - name: Install Dependencies
+        run: poetry install --no-interaction --no-root
+
+      - name: Install ${{ github.repository }}
+        run: |
+          poetry install --no-interaction
+
+      - name: Run Tests
+        run: |
+          poetry install --with test --no-interaction --no-root
+          poetry run pytest
+
+      - name: Build Pure Python Wheels and Source Distribution
+        run: |
+          poetry build
+
+      - name: Check sdist Metadata
+        run: pipx run twine check dist/*
+
+      - name: Store Wheel and sdist Artifacts
+        uses: actions/upload-artifact@v3
+        with:
+          name: distributions
+          path: dist/*
+
+      - name: Upload to Coveralls
+        uses: coverallsapp/github-action@master
+        with:
+          path-to-lcov: coverage.lcov
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+
+  test_pypi_publish:
+    # Test PyPI publish, requires wheels and source dist (sdist)
+    name: Publish ${{ github.repository }} to TestPyPI
+    if: github.event_name == 'release' && github.event.action == 'published'
+    needs: [build_wheels_sdist]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/download-artifact@v3
+        with:
+          name: distributions
+          path: dist
+
+      - uses: pypa/gh-action-pypi-publish@release/v1.6
+        with:
+          user: __token__
+          password: ${{ secrets.TEST_PYPI_API_TOKEN }}
+          repository_url: https://test.pypi.org/legacy/
+          packages_dir: dist/
+          verbose: true
+
+  pypi_publish:
```

Comment:
What's the difference between these two? What workflow are you imagining for testing and then running the actual pypi?

Reply:
For every PR, the `build_wheels_sdist` job will always run tests, upload coverage, and build wheels.

For every new release, the workflow will run `build_wheels_sdist` and then if and only if the `build_wheels_sdist` job succeeds, then `test_pypi_publish` will run. Once `test_pypi_publish` runs to succession, it will then upload the wheels to pypi in `pypi_publish`.